oeps/oep-0046-arch-ddd-domains.rst

```diff
@@ -0,0 +1,327 @@
+**********
+
+Domain Driven Design presents a description of the Open edX system that can be used to clarify decisions in the following areas:
+
+- Communication
+  - Should these two components communicate synchronously or asynchronously?
+  - Is this API using the same terminology that my component is?
+- Level of effort
+  - Should we consider buying this component rather than custom building it?
+  - Is this component something an Open edX operator might want to replace?
+
+
+Specification
+*************
+
+The Open edX system consists of the following bounded contexts.
+
+Terminology and Behaviors
+=========================
+Subdomain:
+  A group of related bounded contexts. A Subdomain is **Core** if it is a key differentiator for Open edX, *Supporting** if it has Open edX specifics, but isn't **Core**, or **Generic** if it has no Open edX business context.
+
+  There should be **no** synchronous communication between separate subdomains. Any exceptions to this must be documented in this OEP.
+
+  Business logic should also **not** be distributed between subdomains. For instance, if the Marketing or Purchasing systems need to calculate a discount, it should be based solely on data in their databases, and code in their repositories, rather than requiring api calls to other subdomains.
+
+Bounded Context:
+  A single service inside a Subdomain. Each Bounded Context owns its own data storage, has its own Ubiquitous Language that describes all of the entities owned by that service, and uses Anti-Corruption Layers to translate incoming objects into that language.
+
+  Bounded Contexts within a Subdomain **should** use asynchronous communication with other Bounded Contexts in the same Subdomain, but may elect to use synchronous communication if required.
+
+  Like Subdomains, **core** Bounded Contexts in a Subdomain implement the key Open edX business logic for that Subdomain. **supporting** Contexts have Open edX logic, but are not **core** to their Subdomain, and **generic**  contexts should be replaceable and contain no Open edX business logic.
+
+Ubiquitous Language:
+  The set of terms and definitions that defines the entities and relationships within a Bounded Context. These terms may not have the same definitions across different Bounded Contexts.
+
+Anti-Corruption Layer:
+  A layer around the edge of a Bounded Context that converts entities and information retrieved from other Bounded Contexts in the same Subdomain into the Ubiquitous Language of this Bounded Context.
+
+Overall Behavior Flows
+======================
+
+A **learner** engages with a **Business Subdomain** to find learning content that meets their learning goals. They purchase that learning content via whatever mode is appropriate for that **Business Subdomain** (direct purchase, free enrollment, subscriptions, etc). The **Business Subdomain** requests fulfillment from the **Learner Subdomain** to allow the **learner** access to their selected learning content.
```

Comment:
* I think having "Business Subdomain" (or "Business Subdomains") in the picture would help.
* If it is too much for one picture, maybe add a high-level picture without bounded contexts?